### PR TITLE
LGA-787 - Adding feature test for adding discrimination and education categories

### PR DESCRIPTION
**Pull request by said-moj**
## What does this pull request do?

Adding feature test for adding discrimination and education categories

## Any other changes that would benefit highlighting?

Intentionally left blank.

### CODE REVIEW

File: tests/behave/features/search_legal_advisers.feature [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Feature: Search for Education and Discrimination categories
+
+  Scenario Outline: Find a legal adviser in <category>
+    Given I am a on the fala landing page
+    When I search for a legal adviser with postcode "<postcode>" and category "<category>"
+    Then I should see a list of legal advisers
+    Examples: Search
+        | postcode         | category |
+        | W1A | disc        |
```
Comment (sldblog) [inline tests/behave/features/search_legal_advisers.feature]: Most editors can auto-format `.feature` files to align these table brackets. It's usually worth doing as it avoids subsequent diffs changing unrelated lines.

File: tests/behave/Dockerfile [hunk 2]
```diff
@@ -12,6 +12,6 @@ USER seluser
 COPY requirements.txt ./
 RUN pip3 install --user -r requirements.txt
 
-COPY . .
+COPY --chown=seluser . .
```
Comment (sldblog) [inline tests/behave/Dockerfile]: Why do we need this change?

File: tests/behave/features/environment.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
+import os
 from selenium import webdriver
 from selenium.webdriver.firefox.options import Options
 
 
 def before_feature(context, feature):
     options = Options()
-    options.headless = True
+    options.headless = os.environ.get("SELENIUM_HEADLESS", "True") == "True"
```
Comment (sldblog) [inline tests/behave/features/environment.py]: Consider converting this to `True` (boolean) or an existence check instead of checking equality to `"True"` -- otherwise users have to remember to put `SELENIUM_HEADLESS=True` in this exact format and case.

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1,5 @@
 node_modules/
-reports/
\ No newline at end of file
+reports/
+.idea
```
Comment (sldblog) [inline .gitignore]: Editor-specific ignores are recommended to be [globally ignored](https://help.github.com/en/articles/ignoring-files#create-a-global-gitignore) as each developer may use a different editor.

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,5 @@
...
+env
```
Comment (sldblog) [inline .gitignore]: Should this be `tests/behave/env/`?

File: tests/behave/features/steps/legal_adviser.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+def step_fala_landing_page(context):
+    get_fala_url(context, "/")
+
+    WebDriverWait(context.driver, 10).until(EC.presence_of_element_located((By.ID, "content")))
+
+
+@when(u'I search for a legal adviser with postcode "{postcode}" and category "{category}"')
+def step_search_legal_adviser(context, postcode, category):
+    context.driver.find_element_by_xpath("//input[@type='text'][@name='postcode']").send_keys(postcode)
+    context.driver.find_element_by_xpath("//input[@name='categories'][@value='{}']".format(category))
```
Comment (sldblog) [inline tests/behave/features/steps/legal_adviser.py]: We don't do anything with the "category" element after finding it, is that intentional?

File: tests/behave/features/steps/legal_adviser.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    context.driver.find_element_by_id("searchButton").click()
+    WebDriverWait(context.driver, 10).until(EC.presence_of_element_located((By.CLASS_NAME, "legal-adviser-results")))
+
+
+@then(u'I should see a list of legal advisers')
+def step_impl(context):
+    if not len(context.driver.find_elements_by_xpath("//div[@class='search-results-list']/ul[@class='org-list']/li")):
```
Comment (sldblog) [inline tests/behave/features/steps/legal_adviser.py]: What do you think about extracting the `len(...)` contents into a nicely named variable? It would help us understand quicker what is going on: `if not len(results_of_the_search)` would read much nicer and would avoid the need to understand the XPath in detail.

File: tests/behave/features/steps/legal_adviser.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        raise Exception("Could not find list of legal advisers")
```
Comment (sldblog) [inline tests/behave/features/steps/legal_adviser.py]: :raising_hand_man: In feature tests, it's typically useful to use assertions like in unit tests; it gives a richer context to the failure and usually conveys the intent better.

```python
results_of_the_search = context.driver.find_elements_by_xpath("//div[@class='search-results-list']/ul[@class='org-list']/li")
assert_greater(results_of_the_search, 0)
```

The other test is using `assert_equal` for this effect (the assertion fail would print out the diff).

File: tests/behave/features/search_legal_advisers.feature [hunk 9]
```diff
@@ -0,0 +1,10 @@
+Feature: Search for Education and Discrimination categories
+
+  Scenario Outline: Find a legal adviser in <category>
+    Given I am a on the fala landing page
```
Comment (sldblog) [inline tests/behave/features/search_legal_advisers.feature]:
```suggestion
    Given I am a user searching for legal advisers
```

File: tests/behave/features/search_legal_advisers.feature [hunk 10]
```diff
@@ -0,0 +1,10 @@
+Feature: Search for Education and Discrimination categories
+
+  Scenario Outline: Find a legal adviser in <category>
```
Comment (sldblog) [inline tests/behave/features/search_legal_advisers.feature]:
```suggestion
  Scenario Outline: Finding legal advisers for specific categories of law
```

File: tests/behave/features/search_legal_advisers.feature [hunk 11]
```diff
@@ -0,0 +1,10 @@
+Feature: Search for Education and Discrimination categories
```
Comment (sldblog) [inline tests/behave/features/search_legal_advisers.feature]: :raising_hand_man: This title should reflect the filename: `search_legal_advisers.feature`

File: tests/behave/Dockerfile (hunk 2 above)
Comment (said-moj) [reply tests/behave/Dockerfile]: The geckodriver.log file seems to get owned by root:root when building the Dockerfile alone.

File: tests/behave/features/steps/legal_adviser.py (hunk 6 above)
Comment (said-moj) [reply tests/behave/features/steps/legal_adviser.py]: Good catch. Turns out I wasn't even filtering by category 🤦‍♂

File: tests/behave/Dockerfile (hunk 2 above)
Comment (said-moj) [reply tests/behave/Dockerfile]: After speaking to @sldblog offline I have decided to remove `--chown=seluser ` addition as it's not supported in all  docker versions

File: tests/behave/features/environment.py [hunk 12]
```diff
@@ -1,11 +1,13 @@
+import os
 from selenium import webdriver
 from selenium.webdriver.firefox.options import Options
 
 
 def before_feature(context, feature):
     options = Options()
-    options.headless = True
+    options.headless = bool(os.environ.get("SELENIUM_HEADLESS", "True") == "True")
```
Comment (MatMoore) [inline tests/behave/features/environment.py]: This and the FALA_URL environment variable seems like a separate change? Is there some additional context I'm missing here? If so, probably worth adding something to the commit message or moving to another commit.

File: tests/behave/features/steps/legal_adviser.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    context.driver.find_element_by_xpath("//input[@name='categories'][@value='{}']".format(category)).click()
+    context.driver.find_element_by_id("searchButton").click()
+    WebDriverWait(context.driver, 10).until(EC.presence_of_element_located((By.CLASS_NAME, "legal-adviser-results")))
+    assert_in("postcode={}".format(postcode), context.driver.current_url)
+    assert_in("categories={}".format(category), context.driver.current_url)
+
+
+@then(u'I should see a list of legal advisers')
+def step_see_list_of_legal_advisers(context):
+    results_of_the_search = context.driver.find_elements_by_xpath("//div[@class='search-results-list']/ul[@class='org-list']/li")
```
Comment (MatMoore) [inline tests/behave/features/steps/legal_adviser.py]: We have been doing something similar lately using Cypress and in their documentation they recommend not relying so much on CSS class names for selecting elements since these are likely to change without changing the functionality of the app. 

What they suggest doing instead is adding data attributes (e.g. `data-test="blablabla"`) that you can use just for the tests.

It might be useful here as well because the advice is not specific to the framework.
The link is https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements

File: tests/behave/features/environment.py (hunk 12 above)
Comment (said-moj) [reply tests/behave/features/environment.py]: The FALA_URL is the url that these test would run against. This is set set in the docker-compose.yml file https://github.com/ministryofjustice/laa-cla-e2e-tests/pull/71/files#diff-eb672b1ddd3aa7dae5f2577f38daa271R128

`options.headless = bool(os.environ.get("SELENIUM_HEADLESS", "True") == "True")`
I added this change so that when debugging / writing test locally you can install selenium locally and see the browser interactions

File: tests/behave/features/steps/legal_adviser.py (hunk 13 above)
Comment (said-moj) [reply tests/behave/features/steps/legal_adviser.py]: I really like this idea and will keep this in mind for future tests

**Comment by sldblog** [comment]
Approved, but would love to wait for @MatMoore as well.

**Comment by sldblog** [comment]
:shipit:
